Add migration guide for pre-extension starter pack (#342)
canonical/sphinx-docs-starter-pack: docs/content/migrate-from-pre-extension.rst

To ensure a smooth migration, update your documentation project to use the last pre-extension version of the Sphinx Documentation Starter Pack. This update ensures that your project is using the latest features and configurations available, minimising the changes required during the migration.

You can find the release tag and branch for this version in the following links:

* `pre-extension branch <https://github.com/canonical/sphinx-docs-starter-pack/blob/pre-extension>`_
* `pre-extension release tag <https://github.com/canonical/sphinx-docs-starter-pack/releases/tag/pre-extension>`_

Set up a new project
--------------------

1. Back up all existing files in your original documentation project. For example, you can rename the original ``docs/`` folder to ``docs_backup/``.

   .. warning::

      If you proceed in the same directory, the following steps will overwrite some of the configuration files in the original project.

2. Follow the steps in the :ref:`initial-setup` guide to initialise an empty project with the extension-based starter pack, at the original file path.

3. Ensure the following files are at the root of your repository:

   - ``.github/workflows/``

>    - ``.wokeignore``

Reviewer: No longer needed as the starter pack has replaced woke with Vale...

Which means I should have removed that ignore file. Whoops.

Author: Removed this line, and also added a comment stating that it has been replaced by Vale in the reference list at the end of file

4. Ensure the following files are moved to their original paths in the project. These files are defaulted to the repository root, but may have be changed upon project needs: 

   - ``.gitignore``
   - ``.readthedocs.yml``

5. Validate the project setup locally by running ``make run`` in the new project directory.

Migrate source files
--------------------

The documentation starter pack has undergone breaking changes with the introduction of the ``canonical-sphinx`` extension. This section guides you through:

- Configuration file changes
- Extension dependencies
- Documentation source migration

For a complete list of the structural changes, refer to the `directory-structure-change`_ section.

Sphinx configuration
~~~~~~~~~~~~~~~~~~~~~

A significant change in the new starter pack is the organisation of the configuration files, summarised in the following table:

.. list-table::
   :widths: 20 40 40
   :header-rows: 1

   * - Configuration file
     - Pre-extension 
     - Extension-based
   * - ``conf.py``
     - Common configurations shared by all starter pack projects
     - Project-specific configurations
   * - ``custom_conf.py``

>      - Project-specific configurations in addition

Reviewer: ```suggestion
     - Project-specific configuration
```

     - Merged into ``conf.py`` and removed

> In the new starter pack, the common configurations are provided by the extension and are loaded automatically when building the documentation. ``docs/conf.py`` is the main and only configuration file that includes project-specific configurations.

Reviewer: ```suggestion
In the new starter pack, many common configurations are provided by the extension and are loaded automatically when building the documentation. ``docs/conf.py`` is the only configuration file, and it contains all project-specific configuration. Sensible defaults are set for general configuration by inclusion of the `canonical-sphinx` extension.
```

Ensure that all the previous changes in the original ``custom_conf.py`` file are copied to the new ``conf.py`` file.  

Dependencies
~~~~~~~~~~~~

If your project requires additional extensions beyond the default list, add the extension list to the new project in ``docs/.sphinx/requirements.txt``.

Documentation source files
~~~~~~~~~~~~~~~~~~~~~~~~~~

> 1. Remove the example documentation files in the starter pack under ``docs/content/``.

Reviewer: ```suggestion
1. Remove the starter pack's documentation files (``index.rst`` and any files in the ``docs/content/*`` dir).
```

> 2. Copy all documentation source files from your original project to the new directory, keeping their original structure. These file may include but are not limited to:

Reviewer: ```suggestion
2. Copy all documentation source files from your original project to the new project, keeping their original structure. These file may include but are not limited to:
```

   - ``.md`` 
   - ``.rst`` 
   - ``.txt``
   - ``.json``
   - images
   - scripts

3. Validate the migration by running ``make run``.

Apply customisation
-------------------

> If your projects have custom configurations or styles, ensure that you identify and  apply these changes to the new documentation project. The extension provides a mechanism to overwrite the default settings with customised resources.

Reviewer: ```suggestion
If your projects have custom configurations or styles, ensure that you identify and apply these changes to the new documentation project.
```

No need to directly support complex customised projects.

2. Create a new CSS file under ``docs/.sphinx/_static``. You can choose any other file location in the project directory, but it's recommended to keep the file structure similar to the original project.
3. Copy the additions and changes to the new empty file.
4. In the ``conf.py``, add the new files into the pre-defined ``html_css_files`` list variable to overwrite the default settings.
5. Build the documentation to verify that the customised styling is applied correctly.

.. _directory-structure-change:

Directory structure changes
----------------------------

After you migrate to the extension, some directories and files are either deleted from the project or moved to a new location.

Assuming that all previous documentation files were in the ``docs/`` sub-directory, the following list illustrates the changes in the directory structure after the migration.

.. code-block:: text

    .
    ├── .github
    │   └── workflows

>     │       ├── automatic-doc-checks.yml
>     │       └── sphinx-python-dependency-build-checks.yml

Reviewer: ```suggestion
    │       ├── automatic-doc-checks.yml
    │       └── markdown-style-checks.yml
```

Don't need dependency build checks unless you really want them, and then it should be run in the starter pack and validated there.
